components/core/src/clp_s/JsonFileIterator.cpp

```diff
@@ -1,74 +1,94 @@
+            size_t new_buf_size = m_buf_size * 2;
+            char* new_buf = new char[new_buf_size + simdjson::SIMDJSON_PADDING];
+            memcpy(new_buf, m_buf, m_buf_size);
+            delete[] m_buf;
+            m_buf = new_buf;
+            m_buf_size = new_buf_size;
+        } else if (m_truncated_bytes > 0) {
+            // move bytes to start of buffer
+            memmove(m_buf, m_buf + (m_buf_occupied - m_truncated_bytes), m_truncated_bytes);
+            m_buf_occupied = m_truncated_bytes;
+        } else {
+            m_buf_occupied = 0;
+        }
 
-    size_t size_read = 0;
-    auto error = m_reader.try_read(m_buf + m_buf_occupied, m_buf_size - m_buf_occupied, size_read);
-    m_buf_occupied += size_read;
-    m_bytes_read += size_read;
+        m_truncated_bytes = 0;
+        m_next_document_position = 0;
 
-    if (error != ErrorCodeSuccess) {
-        m_eof = true;
-    }
+        size_t size_read = 0;
+        auto file_error
+                = m_reader.try_read(m_buf + m_buf_occupied, m_buf_size - m_buf_occupied, size_read);
+        m_buf_occupied += size_read;
+        m_bytes_read += size_read;
 
-    m_parser.iterate_many(
-                    m_buf,
-                    /* length of data */ m_buf_occupied,
-                    /* batch size of data to parse*/ m_buf_occupied
-    )
-            .get(m_stream);
-
-    m_doc_it = m_stream.begin();
-    // only implements != so this is equivalent to
-    // if no json available from buffer and we haven't hit eof
-    // then retry reading the json with a larger buffer up to eof
-    if (false == (m_doc_it != m_stream.end()) && false == m_eof) {
-        read_new_json(m_stream.truncated_bytes());
-    }
+        if (ErrorCodeEndOfFile == file_error) {
+            m_eof = true;
+        } else if (ErrorCodeSuccess != file_error) {
+            m_error_code = simdjson::error_code::IO_ERROR;
+            return false;
+        }
+
+        auto error = m_parser.iterate_many(
+                                     m_buf,
+                                     /* length of data */ m_buf_occupied,
+                                     /* batch size of data to parse*/ m_buf_occupied
+        )
+                             .get(m_stream);
+
+        if (error) {
+            m_error_code = error;
+            return false;
+        }
+
+        m_doc_it = m_stream.begin();
+        // only implements != so this is equivalent to "if no JSON is available from the buffer"
+        if (false == (m_doc_it != m_stream.end())) {
+            m_truncated_bytes = m_buf_occupied;
+        } else {
+            return true;
+        }
+    } while (false == m_eof);
+    return true;
+}
+
+size_t JsonFileIterator::skip_whitespace_and_get_truncated_bytes() {
+    for (;
+         m_next_document_position < m_buf_occupied && std::isspace(m_buf[m_next_document_position]);
+         ++m_next_document_position)
+    {}
```

Comment:
Do you think it's better to change it to
```suggestion
    while (m_next_document_position < m_buf_occupied
           && std::isspace(m_buf[m_next_document_position]))
    {
        ++m_next_document_position;
    }
```